elfi/methods/empirical_density.py

```diff
@@ -0,0 +1,179 @@
+import numpy as np
+import scipy.stats as ss
+from scipy.interpolate import interp1d
+
+
+__all__ = ('EmpiricalDensity', 'ecdf', 'ppf')
+
+
+def ecdf(samples):
+    """Compute an empirical cdf.
+
+    Parameters
+    ----------
+    samples : array_like
+      a univariate sample
+
+    Returns
+    -------
+    empirical_cdf
+      an interpolated function for the estimated cdf
+    """
+    x, y = _handle_endpoints(*_ecdf(samples))
+    return _interp_ecdf(x, y)
+
+
+def ppf(samples):
+    """Compute an empirical quantile function.
+
+    Parameters
+    ----------
+    samples : array_like
+      a univariate sample
+
+    Returns
+    -------
+    empirical_ppf
+      an interpolated function for the estimated quantile function
+    """
+    x, y = _handle_endpoints(*_ecdf(samples))
+    return _interp_ppf(x, y)
+
+
+def _ecdf(x):
+    xs = np.sort(x)
+    ys = np.arange(1, len(xs) +  1)/float(len(xs))
+    return xs, ys
+
+
+def _low(xs, ys):
+    """Compute the intercetion point (x, 0)."""
+    slope = (ys[1] - ys[0])/(xs[1] - xs[0])
+    intersect = -ys[0]/slope + xs[0]
+    return intersect
+
+
+def _high(xs, ys):
+    """Compute the interception point (x, 1)."""
+    slope = (ys[-1] - ys[-2])/(xs[-1] - xs[-2])
+    intersect = (1 - ys[-1])/slope + xs[-1]
+    return intersect
+
+
+def _handle_endpoints(xs, ys):
+    high = _high(xs, ys)
+    low = _low(xs, ys)
+
+    # add endpoints
+    x = np.append(np.insert(xs, 0, low), high)
+    y = np.append(np.insert(ys, 0, 0.), 1.)
+    return x, y
+
+
+def _interp_ecdf(x, y, **kwargs):
+    low, high = x[0], x[-1]
+
+    # linear interpolation
+    f = interp1d(x, y, **kwargs)
+
+    def interp(q):
+        if isinstance(q, np.ndarray):
+            too_low = sum(q < low)
+            too_high = sum(q > high)
+            return np.concatenate([np.zeros(too_low),
+                                   f(q[(q >= low) & (q <= high)]),
+                                   np.ones(too_high)])
+        else:
+            return _scalar_cdf(f, low, high, q)
+
+    return interp
+
+
+def _scalar_cdf(f, low, high, q):
+    if q < low:
+        return 0
+    elif q > high:
+        return 1
+    else:
+        return f(q)
+
+
+def _interp_ppf(x, y, **kwargs):
+    f = interp1d(y, x, **kwargs)
+
+    def interp(p):
+        try:
+            return f(p)
+        except ValueError:
+            raise ValueError("The quantile function is not defined outside [0, 1].")
+
+    return interp
+
+
+class EmpiricalDensity(object):
+    """An empirical approximation of a random variable.
+
+    The density function is approximated using the gaussian
+    kernel density estimation from scipy (scipy.stats.gaussian_kde).
```

Comment:
It still doesn't quite provide the same functionality. `Statsmodels` seems to only provide a step function approximation to the empirical CDF. This code provides a scipy distribution like object with sampling and a pdf. The approximation is also a linear interpolation instead of a step function. Including `statsmodels` would bring a lot of code that is not really relevant for us like interpreting regression model formulas.